services/web-server/src/utils/decode.js

```diff
@@ -0,0 +1,4 @@
+// Decoding:
+// 1. Replace ! with %
+// 2. Decode with decodeUriComponent (which converts %21 to !)
+module.exports = str => decodeURIComponent(str.replace(/!/g, '%'));
```

Comment:
This and encode should probably be in a single module, with some description of the encoding they use beyond "encode" and "decode" -- those names are pretty generic!